### PR TITLE
Swarm<T>.TransferRecentStates became not to send whole states.

### PR DESCRIPTION
**Pull request by limebell**
Instead, it looks up some range of block indexes and send.

### CODE REVIEW

File: Libplanet/Net/Swarm.cs [hunk 1]
```diff
@@ -1238,111 +1240,138 @@ bool render
...
+                        );
+                        break;
                     }
+                }
 
-                    _logger.Debug(
-                        "Finished to store received state refs and block states.");
+                if (offset == -1)
+                {
+                    _logger.Warning("Received states from trusted peers.");
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]: This message seems normal case's, right?

File: Libplanet/Net/Swarm.cs [hunk 2]
```diff
@@ -2003,21 +2033,44 @@ private void TransferRecentStates(GetRecentStates getRecentStates)
...
+                        (@base is HashDigest<SHA256> bbh &&
+                         _store.GetBlockIndex(bbh) is long bbIdx)
+                            ? bbIdx + getRecentStates.Offset
+                            : 0 + getRecentStates.Offset;
+                    long targetIndex =
+                        (target is HashDigest<SHA256> tgt &&
+                         _store.GetBlockIndex(tgt) is long tgtIdx)
+                            ? tgtIdx
+                            : long.MaxValue;
+                    var highestIndex = lowestIndex + FindNextStatesChunkSize - 1 > targetIndex
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]:
```suggestion
                    long highestIndex = lowestIndex + FindNextStatesChunkSize - 1 > targetIndex
```

File: Libplanet/Net/Swarm.cs [hunk 3]
```diff
@@ -2003,21 +2033,44 @@ private void TransferRecentStates(GetRecentStates getRecentStates)
...
                         .Select(refs => refs.Last())
                         .ToImmutableHashSet()
+                        .Where(bh => !(_store.GetBlockStates(bh) is null))
                         .Select(bh =>
                             new KeyValuePair<
                                 HashDigest<SHA256>,
                                 IImmutableDictionary<Address, IValue>
-                            >(bh, _store.GetBlockStates(bh))
-                        )
+                            >(bh, _store.GetBlockStates(bh)))
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]: I guess it's better if we call `_store.GetBlockStates()` once and filter by its value.

File: Libplanet/Net/Swarm.cs (hunk 1 above)
Comment (limebell) [reply Libplanet/Net/Swarm.cs]: What do you mean by normal case?

Comment (longfin) [reply Libplanet/Net/Swarm.cs]: `"Received states from trusted peers."` doesn't seem a warning message. 😕

Comment (limebell) [reply Libplanet/Net/Swarm.cs]: You are right. I'll fix it :)

File: Libplanet/Net/Swarm.cs [hunk 4]
```diff
@@ -2003,21 +2033,43 @@ private void TransferRecentStates(GetRecentStates getRecentStates)
...
+                    nextOffset = highestIndex == targetIndex
+                        ? -1
+                        : getRecentStates.Offset + FindNextStatesChunkSize;
+
                     stateRefs = _store.ListAllStateReferences(
                         chainId,
-                        onlyAfter: @base,
-                        ignoreAfter: target
+                        lowestIndex: lowestIndex,
+                        highestIndex: highestIndex
```
Comment (earlbread) [inline Libplanet/Net/Swarm.cs]: How about leaving only one API for `ListAllStateReferences`?
If we use the previous one this code can be
```
                    var lowestBlockHash = lowestIndex == 0
                        ? null
                        : _store.IndexBlockHash(chainId, lowestIndex - 1);
                    var highestBlockHash = _store.IndexBlockHash(chainId, highestIndex);

                    stateRefs = _store.ListAllStateReferences(
                        chainId,
                        lowestBlockHash,
                        highestBlockHash
                    );
```
Or we can remove the previous API.

File: CHANGES.md [hunk 5]
```diff
@@ -34,17 +34,26 @@ To be released.
 
 ### Backward-incompatible network protocol changes
 
+ -  Added long typed `offset` parameter to `RecentStates` and
+    `GetRecentStates` messages.  [[#703]]
+ -  Added int typed `iteration` parameter to `RecentStates` message.
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion
 -  Added `int`-typed `iteration` parameter to `RecentStates` message.
```

File: Libplanet/Net/Swarm.cs [hunk 6]
```diff
@@ -1242,111 +1244,132 @@ bool render
...
+                        // when synchronizing previous states.
+                        _logger.Debug("Received recent states from a peer ({Peer}).", peer);
+                    }
+                    catch (TimeoutException e)
+                    {
+                        _logger.Error(
+                            "Failed to receive recent states from a peer ({Peer}): " + e,
+                            peer
+                        );
+                        continue;
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]: We should `break` here to exit while loop.

File: CHANGES.md [hunk 7]
```diff
@@ -31,9 +31,18 @@ To be released.
  -  Removed `StoreExtension.ListAllStateReferences<T>()` method.  [[#701]]
  -  Added the `genesisBlock` parameter to
     `BlockChain<T>()` constructor.  [[#688]]
+ -  Removed `StateReferenceDownloadState`.  [[#703]]
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion
 -  Removed `StateReferenceDownloadState` class.  [[#703]]
```

File: CHANGES.md [hunk 8]
```diff
@@ -31,9 +31,18 @@ To be released.
...
+ -  Removed `BlockStateDownloadState`.  [[#703]]
```
Comment (dahlia) [inline CHANGES.md]:
```suggestion
 -  Removed `BlockStateDownloadState` class.  [[#703]]
```

File: CHANGES.md [hunk 9]
```diff
@@ -31,9 +31,18 @@ To be released.
...
+ -  Added `StateDownloadState` which reports state preloading iteration
```
Comment (dahlia) [inline CHANGES.md]: This one should go to the added APIs.

```suggestion
 -  Added `StateDownloadState` class which reports state preloading iteration
```

File: Libplanet.Tests/Net/Protocols/TestSwarm.cs [hunk 10]
```diff
@@ -53,6 +53,8 @@ public class TestSwarm : ISwarm
             _random = new Random();
         }
 
+        public IEnumerable<BoundPeer> Peers => Protocol.Peers;
```
Comment (dahlia) [inline Libplanet.Tests/Net/Protocols/TestSwarm.cs]: Should this be necessarily `public`? If so it's better to have a single sentence docs as well.

File: Libplanet/Net/StateDownloadState.cs [hunk 11]
```diff
@@ -3,17 +3,17 @@ namespace Libplanet.Net
     /// <summary>
     /// Indicates a progress of downloading state references.
```
Comment (dahlia) [inline Libplanet/Net/StateDownloadState.cs]: This should be updated too.

File: Libplanet/Net/StateDownloadState.cs [hunk 12]
```diff
@@ -3,17 +3,17 @@ namespace Libplanet.Net
...
     /// </summary>
-    public class StateReferenceDownloadState : PreloadState
+    public class StateDownloadState : PreloadState
     {
         /// <summary>
         /// Total number of state references to receive in the current batch.
```
Comment (dahlia) [inline Libplanet/Net/StateDownloadState.cs]: Outdated term.

File: Libplanet/Net/StateDownloadState.cs [hunk 13]
```diff
@@ -3,17 +3,17 @@ namespace Libplanet.Net
...
         /// </summary>
-        public int TotalStateReferenceCount { get; internal set; }
+        public int TotalIterationCount { get; internal set; }
 
         /// <summary>
         /// The number of received state references until now.
```
Comment (dahlia) [inline Libplanet/Net/StateDownloadState.cs]: Outdated term.

File: Libplanet.Tests/Net/Protocols/TestSwarm.cs (hunk 10 above)
Comment (limebell) [reply Libplanet.Tests/Net/Protocols/TestSwarm.cs]: I made it public since `TestSwarm` is a test class which in `Libplanet.Tests`.

**Pull request merged**